src/package-sources/npmjs/npm-js-follower.lambda.ts

```diff
@@ -226,107 +241,180 @@ export async function handler(event: ScheduledEvent, context: Context) {
-    );
-    let data = JSON.parse(body, (key, value) => {
-      if (key !== 'knownVersions') {
-        return value;
-      }
-      const map = new Map<string, Date>();
-      for (const [pkgVersion, iso] of Object.entries(value)) {
-        if (typeof iso === 'string' || typeof iso === 'number') {
-          map.set(pkgVersion, new Date(iso));
-        } else {
-          console.error(`Ignoring invalid entry: ${pkgVersion} => ${iso}`);
-        }
-      }
-      return map;
-    });
-    if (typeof data === 'number') {
-      data = { marker: data.toFixed(), knownVersions: new Map() };
-    }
-    console.log(`Read last transaction marker: ${data.marker}`);
-
-    const dbUpdateSeq = (await registry.info()).update_seq;
-    if (dbUpdateSeq < data.marker) {
-      console.warn(
-        `Current DB update_seq (${dbUpdateSeq}) is lower than marker (CouchDB instance was likely replaced), resetting to 0!`
-      );
-      return { marker: '0', knownVersions: data.knownVersions };
+      throw new Error(`Response Body for ${key} is empty`);
     }
-
-    return data;
+    return await decompressContent(response.Body, response.ContentEncoding);
   } catch (error: any) {
     if (error instanceof NoSuchKey || error.name === 'NoSuchKey') {
-      console.warn(
-        `Marker object (s3://${stagingBucket}/${MARKER_FILE_NAME}) does not exist, starting from scratch`
-      );
-      return { marker: '0', knownVersions: new Map() };
+      console.warn(warningMessage);
+      return null;
     }
     // re-throw unexpected errors
     throw error;
   }
 }
 
+/**
+ * Loads the last known versions from S3.
+ *
+ * @returns the value of the last known versions.
+ */
+async function loadLastKnownVersions(
+  stagingBucket: string
+): Promise<Map<string, Date>> {
+  const warningMessage = `Known versions object (s3://${stagingBucket}/${KNOWN_VERSIONS_FILE_NAME}) does not exist, starting from scratch`;
+  const content = await loadContentFromS3(
+    stagingBucket,
+    KNOWN_VERSIONS_FILE_NAME,
+    warningMessage
+  );
+
+  // Known Versions does not exist, starting from scratch
+  if (content === null) {
+    return new Map<string, Date>();
+  }
+
+  // Known Versions exists, generating map from the data
+  const data: { [key: string]: string } = JSON.parse(content, (_, value) => {
+    if (typeof value === 'string') {
+      const date = new Date(value);
+      if (!isNaN(date.getTime())) return date;
+    }
+    return value;
+  });
+
+  console.log(`Loaded last known versions data`);
+
+  return new Map(Object.entries(data).map(([k, v]) => [k, new Date(v)]));
+}
+
+/**
+ * Loads the last transaction marker from S3.
+ *
+ * @param registry a Nano database corresponding to the Npmjs.com CouchDB instance.
+ *
+ * @returns the value of the last transaction marker.
+ */
+async function loadLastTransactionMarker(
+  stagingBucket: string,
+  registry: CouchChanges
+): Promise<{ marker: string | number; knownVersions?: Map<string, Date> }> {
```

Comment:
This seems like *the* choice function whose job it should be to normalize the wacky outside world into a consistent data model.

```suggestion
): Promise<{ marker: number; knownVersions: Map<string, Date> }> {
```

Reply:
i disagree, but i wrote another function that can act as _the_ choice function which normalizes the wacky outside world.